### PR TITLE
enhance: Support readable JSON file import for Float16/BFloat16/SparseFloat

### PR DESCRIPTION
**Pull request by cydrain**
Issue: #22837

### CODE REVIEW

File: pkg/util/typeutil/schema.go [hunk 1]
```diff
@@ -1519,60 +1536,57 @@ type sparseFloatVectorJSONRepresentation struct {
...
 
-		for idx, val := range vec2 {
-			indices = append(indices, idx)
-			values = append(values, val)
+	// try format1
+	indices, ok1 := input["indices"].([]uint32)
+	values, ok2 := input["values"].([]float32)
+
+	// try format2
+	if !ok1 && !ok2 {
```
Comment (zhengbuqian) [inline pkg/util/typeutil/schema.go:1557]: `if !ok1 || !ok2` ?

File: pkg/util/typeutil/schema.go [hunk 2]
```diff
@@ -1519,60 +1536,57 @@ type sparseFloatVectorJSONRepresentation struct {
...
+		for k, v := range input {
+			idx, err := strconv.ParseUint(k, 0, 32)
+			if err != nil {
+				return nil, err
+			}
+
+			var val float64
```
Comment (zhengbuqian) [inline pkg/util/typeutil/schema.go:1564]: why not `var val float32`?

File: pkg/util/typeutil/schema.go [hunk 3]
```diff
@@ -1519,60 +1536,57 @@ type sparseFloatVectorJSONRepresentation struct {
...
 	}
 
-	indexOrder := make([]int, len(indices))
-	for i := range indexOrder {
-		indexOrder[i] = i
+	if len(indices) != len(values) {
+		return nil, fmt.Errorf("indices and values length mismatch")
 	}
-
-	sort.Slice(indexOrder, func(i, j int) bool {
```
Comment (zhengbuqian) [inline pkg/util/typeutil/schema.go:1564]: let's still sort in this method. `CreateSparseFloatRow` is also used to generate test data that may be inserted by the user.

File: pkg/util/typeutil/schema_test.go [hunk 4]
```diff
@@ -2057,7 +2057,7 @@ func TestValidateSparseFloatRows(t *testing.T) {
 			CreateSparseFloatRow([]uint32{100, 2000, 500}, []float32{1.0, 2.0, 3.0}),
 		}
 		err := ValidateSparseFloatRows(rows...)
-		assert.Error(t, err)
```
Comment (zhengbuqian) [inline pkg/util/typeutil/schema_test.go:2060]: this test is to catch invalid user input, not to test `CreateSparseFloatRow`

File: pkg/util/typeutil/schema_test.go [hunk 5]
```diff
@@ -2120,107 +2120,107 @@ func TestValidateSparseFloatRows(t *testing.T) {
...
 		assert.NoError(t, err)
 		assert.Equal(t, CreateSparseFloatRow([]uint32{1, 3, 5}, []float32{2.0, 1.0, 3.0}), res)
 	})
 
-	t.Run("invalid row 1", func(t *testing.T) {
-		row := []byte(`{"indices":[1,3,5],"values":[1.0,2.0,3.0`)
-		_, err := CreateSparseFloatRowFromJSON(row)
-		assert.Error(t, err)
-	})
+	//t.Run("invalid row 1", func(t *testing.T) {
```
Comment (zhengbuqian) [inline pkg/util/typeutil/schema_test.go:2136]: delete this section

File: pkg/util/typeutil/schema_test.go [hunk 6]
```diff
@@ -2120,107 +2120,107 @@ func TestValidateSparseFloatRows(t *testing.T) {
...
 		assert.Error(t, err)
 	})
 
-	t.Run("invalid row 5", func(t *testing.T) {
-		row := []byte(`{"indices":[-3],"values":[0.2]`)
-		_, err := CreateSparseFloatRowFromJSON(row)
-		assert.Error(t, err)
-	})
+	//t.Run("invalid row 5", func(t *testing.T) {
+	//	row := map[string]interface{}{"indices": []uint32{-3}, "values": []float32{0.2}}
```
Comment (zhengbuqian) [inline pkg/util/typeutil/schema_test.go:2161]: delete this section

File: pkg/util/typeutil/schema_test.go [hunk 7]
```diff
@@ -2120,107 +2120,107 @@ func TestValidateSparseFloatRows(t *testing.T) {
...
 		_, err := CreateSparseFloatRowFromJSON(row)
 		assert.Error(t, err)
 	})
 
-	t.Run("invalid dict row 4", func(t *testing.T) {
-		row := []byte(`{"1": 1.0, "3": 2.0, "5": }`)
-		_, err := CreateSparseFloatRowFromJSON(row)
-		assert.Error(t, err)
-	})
+	//t.Run("invalid dict row 4", func(t *testing.T) {
```
Comment (zhengbuqian) [inline pkg/util/typeutil/schema_test.go:2204]: delete this section

File: pkg/util/typeutil/schema.go [hunk 8]
```diff
@@ -1519,60 +1536,57 @@ type sparseFloatVectorJSONRepresentation struct {
...
+	if len(input) == 0 {
+		return nil, fmt.Errorf("empty JSON input")
+	}
 
-		for idx, val := range vec2 {
-			indices = append(indices, idx)
-			values = append(values, val)
+	// try format1
+	indices, ok1 := input["indices"].([]uint32)
+	values, ok2 := input["values"].([]float32)
```
Comment (bigsheeper) [inline pkg/util/typeutil/schema.go:1554]: what if the input is `{"indices": [1, 2, 3], "1": 0.1, "2": 0.2, "3": 0.3}`, I believe we require a more stringent inspection.

File: pkg/util/typeutil/schema.go [hunk 9]
```diff
@@ -1519,60 +1536,57 @@ type sparseFloatVectorJSONRepresentation struct {
...
+			if err != nil {
+				return nil, err
+			}
+
+			var val float64
+			val, ok := v.(float64)
+			if !ok {
+				num, ok := v.(json.Number)
+				if !ok {
+					return nil, fmt.Errorf("invalid float32 in JSON")
```
Comment (bigsheeper) [inline pkg/util/typeutil/schema.go:1569]: print the type of v as well

File: pkg/util/typeutil/schema.go (hunk 1 above)
Comment (cydrain) [reply pkg/util/typeutil/schema.go:1557]: added "ok1 != ok2" branch

File: pkg/util/typeutil/schema.go (hunk 2 above)
Comment (cydrain) [reply pkg/util/typeutil/schema.go:1564]: json parser always return float64

File: pkg/util/typeutil/schema_test.go (hunk 5 above)
Comment (cydrain) [reply pkg/util/typeutil/schema_test.go:2136]: done

File: pkg/util/typeutil/schema_test.go (hunk 6 above)
Comment (cydrain) [reply pkg/util/typeutil/schema_test.go:2161]: done

File: pkg/util/typeutil/schema_test.go (hunk 7 above)
Comment (cydrain) [reply pkg/util/typeutil/schema_test.go:2204]: done

File: pkg/util/typeutil/schema.go (hunk 8 above)
Comment (cydrain) [reply pkg/util/typeutil/schema.go:1554]: we will return error for this mixed format

File: pkg/util/typeutil/schema.go (hunk 9 above)
Comment (cydrain) [reply pkg/util/typeutil/schema.go:1569]: OK

**Comment by zhengbuqian** [comment]
reviewed sparse part
/lgtm
/approve

File: pkg/util/typeutil/schema.go (hunk 3 above)
Comment (cydrain) [reply pkg/util/typeutil/schema.go:1564]: OK

File: pkg/util/typeutil/schema_test.go (hunk 4 above)
Comment (cydrain) [reply pkg/util/typeutil/schema_test.go]: change it back

**Pull request merged**